cpp/src/gandiva/to_date_functions_holder.h

```diff
@@ -0,0 +1,253 @@
+        "Error parsing value " + std::string(data, data_len) + " for given format.";
+    context->set_error_msg(err_msg.c_str());
+  }
+
+  static Status Make(const FunctionNode& node, std::shared_ptr<HOLDER_TYPE>* holder,
+                     const std::string& function_name) {
+    if (node.children().size() != 2 && node.children().size() != 3) {
+      return Status::Invalid(function_name +
+                             " function requires two or three parameters");
+    }
+
+    auto literal_pattern = dynamic_cast<LiteralNode*>(node.children().at(1).get());
+    if (literal_pattern == NULLPTR) {
+      return Status::Invalid(function_name +
+                             " function requires a literal as the second parameter");
+    }
+
+    auto literal_type = literal_pattern->return_type()->id();
+    if (literal_type != arrow::Type::STRING && literal_type != arrow::Type::BINARY) {
+      return Status::Invalid(
+          function_name + " function requires a string literal as the second parameter");
+    }
+    auto pattern = arrow::util::get<std::string>(literal_pattern->holder());
+
+    int suppress_errors = 0;
+    if (node.children().size() == 3) {
+      auto literal_suppress_errors =
+          dynamic_cast<LiteralNode*>(node.children().at(2).get());
+      if (literal_suppress_errors == NULLPTR) {
+        return Status::Invalid("The (optional) third parameter to " + function_name +
+                               " function needs to be an integer "
+                               "literal to indicate whether to suppress the error");
+      }
+
+      literal_type = literal_suppress_errors->return_type()->id();
+      if (literal_type != arrow::Type::INT32) {
+        return Status::Invalid("The (optional) third parameter to " + function_name +
+                               " function needs to be an integer "
+                               "literal to indicate whether to suppress the error");
+      }
+      suppress_errors = arrow::util::get<int>(literal_suppress_errors->holder());
+    }
+
+    return Make(pattern, suppress_errors, holder);
+  }
+
+  static Status Make(const std::string& sql_pattern, int32_t suppress_errors,
+                     std::shared_ptr<HOLDER_TYPE>* holder) {
+    std::shared_ptr<std::string> transformed_pattern;
+    ARROW_RETURN_NOT_OK(DateUtils::ToInternalFormat(sql_pattern, &transformed_pattern));
+    auto lholder = std::shared_ptr<HOLDER_TYPE>(
+        new HOLDER_TYPE(*(transformed_pattern.get()), suppress_errors));
+    *holder = lholder;
+    return Status::OK();
+  }
+};
+
+/// Function Holder for SQL 'to_date'
+class GANDIVA_EXPORT ToDateHolder : public gandiva::ToDateFunctionsHolder<ToDateHolder> {
+ public:
+  ~ToDateHolder() override = default;
+
+  ToDateHolder(const std::string& pattern, int32_t suppress_errors)
+      : ToDateFunctionsHolder<ToDateHolder>(pattern, suppress_errors, true,
+                                            ::arrow::TimeUnit::SECOND) {}
```

Comment:
Todate should be returning millis instead of seconds.

Reply:
Ok. Seems we are converting it into millis again.

Reply:
I changed to return in millis